### PR TITLE
refactor: use @bitcoinerlab/secp256k1 and @noble/secp256k1

### PR DESCRIPTION
**Pull request by mvayngrib**
~todo: verify correctness 😛~

### CODE REVIEW

**Comment by alexandrius** [comment]
All tests are successful.
<details>
  <summary>Test Results</summary>

```
yarn test
yarn run v1.19.1
$ standard && mocha


  hdkey
    + fromMasterSeed
      ✓ should properly derive the chain path: m
      ✓ should properly derive the chain path: m/0'
      ✓ should properly derive the chain path: m/0'/1
      ✓ should properly derive the chain path: m/0'/1/2'
      ✓ should properly derive the chain path: m/0'/1/2'/2
      ✓ should properly derive the chain path: m/0'/1/2'/2/1000000000
      ✓ should properly derive the chain path: m
      ✓ should properly derive the chain path: m/0
      ✓ should properly derive the chain path: m/0/2147483647'
      ✓ should properly derive the chain path: m/0/2147483647'/1
      ✓ should properly derive the chain path: m/0/2147483647'/1/2147483646'
      ✓ should properly derive the chain path: m/0/2147483647'/1/2147483646'/2
      > m toJSON() / fromJSON()
        ✓ should return an object read for JSON serialization
      > m/0' toJSON() / fromJSON()
        ✓ should return an object read for JSON serialization
      > m/0'/1 toJSON() / fromJSON()
        ✓ should return an object read for JSON serialization
      > m/0'/1/2' toJSON() / fromJSON()
        ✓ should return an object read for JSON serialization
      > m/0'/1/2'/2 toJSON() / fromJSON()
        ✓ should return an object read for JSON serialization
      > m/0'/1/2'/2/1000000000 toJSON() / fromJSON()
        ✓ should return an object read for JSON serialization
      > m toJSON() / fromJSON()
        ✓ should return an object read for JSON serialization
      > m/0 toJSON() / fromJSON()
        ✓ should return an object read for JSON serialization
      > m/0/2147483647' toJSON() / fromJSON()
        ✓ should return an object read for JSON serialization
      > m/0/2147483647'/1 toJSON() / fromJSON()
        ✓ should return an object read for JSON serialization
      > m/0/2147483647'/1/2147483646' toJSON() / fromJSON()
        ✓ should return an object read for JSON serialization
      > m/0/2147483647'/1/2147483646'/2 toJSON() / fromJSON()
        ✓ should return an object read for JSON serialization
    - privateKey
      ✓ should throw an error if incorrect key size
    - publicKey
      ✓ should throw an error if incorrect key size
      ✓ should not throw if key is 33 bytes (compressed)
      ✓ should not throw if key is 65 bytes (not compressed)
    + fromExtendedKey()
      > when private
        ✓ should parse it
      > when public
        ✓ should parse it
        ✓ should parse it without verification
    > when signing
      ✓ should work
    > when deriving public key
      ✓ should work
    > when private key integer is less than 32 bytes
      ✓ should work
    HARDENED_OFFSET
      ✓ should be set
    > when private key has leading zeros
      ✓ will include leading zeros when hashing to derive child
    > when private key is null
      ✓ privateExtendedKey should return null and not throw
     - when the path given to derive contains only the master extended key
      ✓ should return the same hdkey instance
     - when the path given to derive does not begin with master extended key
      ✓ should throw an error
    - after wipePrivateData()
      ✓ should not have private data
      ✓ should have correct data
      ✓ should be able to verify signatures
      ✓ should not throw if called on hdkey without private data
    Deriving a child key does not mutate the internal state
      ✓ should not mutate it when deriving with a private key
      ✓ should not mutate it when deriving without a private key


  45 passing (107ms)

✨  Done in 1.05s.
```
</details>

@mvayngrib Do you think we need to add more test cases?

**Comment by alexandrius** [comment]
The change assumes correctness of `@bitcoinerlab/secp256k1`, HDKey's API is tested public and private keys. We can add some tests there.

I'd add create separate tests for `@exodus/secp256k1` vs `@bitcoinerlab/secp256k1`.

`publicKeyVerify` ~> `isPoint`
`publicKeyConvert` ~> `pointCompress`
`privateKeyTweakAdd` ~> `privateAdd`
`publicKeyTweakAdd` ~> `pointAddScalar`
`ecdsaSign` ~> `sign`
`ecdsaVerify` ~> `verify`
`privateKeyVerify` ~> `isPrivate`
`publicKeyCreate` ~> `pointFromScalar`


Will update more later.

cc @andrejborstnik 

@RyanZim Can you also chime in here?

**Comment by RyanZim** [comment]
In reviewing the tests here, one thing that sticks out is the `publicKey` API. It would be great to have actual data correctness checks added here, to verify that the resulting public key is properly compressed/left compressed: https://github.com/ExodusMovement/hdkey/blob/472ee65ba01534f283d9d747466417f22c0f735e/test/hdkey.test.js#L61-L76

These test additions ideally should be contributed to upstream and backported here.

**Comment by alexandrius** [comment]
Added more fixtures, point compression test.

@mvayngrib Should I mark "Ready for review"?
Also could you update the head?

**Review by headfire94 — APPROVED** [review]
utACK, i rely on unit tests here)

**Comment by mvayngrib** [comment]
> Also could you update the head?

@alexandrius is this PR behind the `exodus` branch?

**Comment by alexandrius** [comment]
@mvayngrib sorry I meant PR description

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@exodus/hdkey",
-  "version": "2.1.0-exodus.0",
+  "version": "2.1.1-exodus.1",
```
Comment (mvayngrib) [inline package.json]: let's not bundle this in with these changes, reverted e1b5ed92ee86fd6fd3eb97867880e5a4b731ec50

File: test/fixtures/hdkey.json [hunk 2]
```diff
@@ -71,9 +71,99 @@
       "path": "m/0/2147483647'/1/2147483646'/2",
       "public": "xpub6FnCn6nSzZAw5Tw7cgR9bi15UV96gLZhjDstkXXxvCLsUXBGXPdSnLFbdpq8p9HmGsApME5hQTZ3emM2rnY5agb9rXpVGyy3bdW6EEgAtqt",
       "private": "xprvA2nrNbFZABcdryreWet9Ea4LvTJcGsqrMzxHx98MMrotbir7yrKCEXw7nadnHM8Dq38EGfSh6dqA9QWTyefMLEcBYJUuekgW4BYPJcr9E7j"
+    },
+    {
+      "seed": "69afbf0608755b3480ca7314c145695c64f973c988790eabb165d19809c7991acecfe4518b1456f274e171d496ddd044942278577b6efcdb69a6374d5341ea0a",
```
Comment (mvayngrib) [inline test/fixtures/hdkey.json:76]: @alexandrius what did u use to generate these?

**Review by RyanZim — APPROVED** [review]
Quick utACK; did not verify correctness.

Can we get these new tests contributed to upstream?

**Comment by mvayngrib** [comment]
> I'd add create separate tests for @exodus/secp256k1 vs @bitcoinerlab/secp256k1.

> publicKeyVerify ~> isPoint
> publicKeyConvert ~> pointCompress
> privateKeyTweakAdd ~> privateAdd
> publicKeyTweakAdd ~> pointAddScalar
> ecdsaSign ~> sign
> ecdsaVerify ~> verify
> privateKeyVerify ~> isPrivate
> publicKeyCreate ~> pointFromScalar

comparing the two libs, they all look correct to me (the "tweak" methods are in-place vs immutable, but the way we use them it doesn't matter). @alexandrius did u still want to add unit tests for these?

**Comment by alexandrius** [comment]
@mvayngrib I don't think we need anything else, it was quick initial thought

**Comment by alexandrius** [comment]
@RyanZim will do

**Comment by mvayngrib** [comment]
> @mvayngrib I don't think we need anything else, it was quick initial thought

for something this core/sensitive, i think u had the right idea when u said to add them 😅

File: test/fixtures/hdkey.json (hunk 2 above)
Comment (alexandrius) [reply test/fixtures/hdkey.json:76]: Extracted from:
`src/_local_modules/keys/derive.js` `createKeyAccess`

**Comment by alexandrius** [comment]
>for something this core/sensitive, i think u had the right idea when u said to add them 😅

I can backport tests to previous version. I did on my end was successful

**Comment by alexandrius** [comment]
Should we go ahead with merge and publish?

**Comment by alexandrius** [comment]
@RyanZim BTW created pull request to upstream
https://github.com/cryptocoinjs/hdkey/pull/60

**Comment by mvayngrib** [comment]
> I can backport tests to previous version. I did on my end was successful

sorry, not following. I mean can we add some unit tests for the underlying functions u listed here https://github.com/ExodusMovement/hdkey/pull/1#issuecomment-1841141586

**Comment by alexandrius** [comment]
@mvayngrib I can but I don't think we should do that since we also have this: https://github.com/ExodusMovement/secp256k1-node/pull/1

**Comment by mvayngrib** [comment]
ah, so the unit tests there will make sure those methods match correctly? sgtm

**Comment by mvayngrib** [comment]
@alexandrius sorry, maybe i'm getting confused. With that PR, why do we even need this one?

**Comment by alexandrius** [comment]
@mvayngrib when this PR was created I think we thought only this PR would be enough to match perf. Another point is to incrementally jump on noble

**Comment by mvayngrib** [comment]
could u walk me through the plan, is it this?
1. finish unit tests in that PR to make us comfortable with this PR
2. merge + release + use this
3. sometime later, revert this and use that PR instead?

**Comment by alexandrius** [comment]
@mvayngrib basically yes. Except swap 1st and 2nd steps. We have all tests to check hdkey correctness comparing the methods above do not belong in here

**Comment by mvayngrib** [comment]
as long as we don't ship this to users before we have that unit test suite, i'm okay with it

**Comment by alexandrius** [comment]
@mvayngrib without that PR we won't be able to ship to users anyway. 
However I would argue testing those methods do not belong here anyway since we have hdkey correctness checks

**Pull request merged**